### PR TITLE
A few more data structures to the code index

### PR DESCRIPTION
**Pull request by mark-i-m**
cc #50

### CODE REVIEW

File: src/appendix/code-index.md [hunk 1]
```diff
@@ -6,24 +6,38 @@ compiler.
...
-`Session` | struct | The data associated with a compilation session | [the Parser], [The Rustc Driver] | [src/librustc/session/mod.html](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/session/struct.Session.html)
+`NodeId` | struct | One of four types of HIR node identifiers. Being phased out. | [Identifiers in the HIR] | [src/libsyntax/ast.rs](https://doc.rust-lang.org/nightly/nightly-rustc/syntax/ast/struct.NodeId.html)
+`ParamEnv` | struct | Information about generic parameters or `Self`, useful for working with associated or generic items | [Parameter Environment] | [src/librustc/ty/mod.rs](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/ty/struct.ParamEnv.html)
+`ParseSess` | struct | This struct contains information about a parsing session | [The parser] | [src/libsyntax/parse/mod.rs](https://doc.rust-lang.org/nightly/nightly-rustc/syntax/parse/struct.ParseSess.html)
+`Rib` | struct | Represents a single scope of names | [Name resolution] | [src/librustc_resolve/lib.rs](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_resolve/struct.Rib.html)
```
Comment (nikomatsakis) [inline src/appendix/code-index.md]: I think this will conflict with some other PR I just merged :)

File: src/appendix/code-index.md [hunk 2]
```diff
@@ -6,24 +6,38 @@ compiler.
...
 `StringReader` | struct | This is the lexer used during parsing. It consumes characters from the raw source code being compiled and produces a series of tokens for use by the rest of the parser | [The parser] |  [src/libsyntax/parse/lexer/mod.rs](https://doc.rust-lang.org/nightly/nightly-rustc/syntax/parse/lexer/struct.StringReader.html)
+`syntax::token_stream::TokenStream` | struct | An abstract sequence of tokens, organized into `TokenTree`s | [The parser], [Macro expansion] | [src/libsyntax/tokenstream.rs](https://doc.rust-lang.org/nightly/nightly-rustc/syntax/tokenstream/struct.TokenStream.html)
 `TraitDef` | struct | This struct contains a trait's definition with type information | [The `ty` modules] |  [src/librustc/ty/trait_def.rs](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/ty/trait_def/struct.TraitDef.html)
+`TraitRef` | struct | Information about a trait reference (e.g. `P0: Trait<P1...Pn>`) | [Trait Solving: Goals and Clauses], [Trait Solving: Lowering impls]  |  [src/librustc/ty/sty.rs](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/ty/struct.TraitRef.html)
```
Comment (nikomatsakis) [inline src/appendix/code-index.md]: How about: "the combination of a trait and its input types" or something like that?

File: src/appendix/code-index.md (hunk 1 above)
Comment (nikomatsakis) [reply src/appendix/code-index.md]: Oh, I see, https://github.com/rust-lang-nursery/rustc-guide/pull/152/ was editing the *glossary*

File: src/appendix/code-index.md (hunk 2 above)
Comment (mark-i-m) [reply src/appendix/code-index.md]: Done

**Pull request merged**